lib/ruby_indexer/lib/ruby_indexer/entry.rb

```diff
@@ -617,15 +632,56 @@ def matches?(arguments)
             max_pos = Float::INFINITY
           when ForwardingParameter
             max_pos = Float::INFINITY
+            has_forward = true
+          when KeywordParameter, OptionalKeywordParameter
+            names << param.name
+          when KeywordRestParameter
+            has_keyword_rest = true
           end
         end
 
-        _keyword_hash_node, positional_args = arguments.partition { |arg| arg.is_a?(Prism::KeywordHashNode) }
-        argument_length_is_unknown = positional_args.any? do |arg|
-          arg.is_a?(Prism::SplatNode) || arg.is_a?(Prism::ForwardingArgumentsNode)
+        keyword_hash_nodes, positional_args = arguments.partition { |arg| arg.is_a?(Prism::KeywordHashNode) }
+        keyword_args = T.cast(keyword_hash_nodes.first, T.nilable(Prism::KeywordHashNode))&.elements
+        forwarding_arguments, positionals = positional_args.partition do |arg|
+          arg.is_a?(Prism::ForwardingArgumentsNode)
         end
 
-        argument_length_is_unknown || (min_pos..max_pos).cover?(positional_args.length)
+        (has_forward && min_pos == 0) ||
+          (positionals.empty? && forwarding_arguments.any?) ||
+          (
+            positional_arguments_match?(positionals, forwarding_arguments, keyword_args, min_pos, max_pos) &&
+            (has_forward || has_keyword_rest || keyword_arguments_match?(keyword_args, names))
+          )
+      end
+
+      sig do
+        params(
+          positional_args: T::Array[Prism::Node],
+          forwarding_arguments: T::Array[Prism::Node],
+          keyword_args: T.nilable(T::Array[Prism::Node]),
+          min_pos: Integer,
+          max_pos: T.any(Integer, Float),
+        ).returns(T::Boolean)
+      end
+      def positional_arguments_match?(positional_args, forwarding_arguments, keyword_args, min_pos, max_pos)
+        # If the method accepts at least one positional argument and a splat has been passed
```

Comment:
I wanted to avoid paying the price of computing all variables every time. We lose the ability to short circuit if we set everything to a variable.

Do you find that the comments are confusing? Or do you just prefer the variables pattern?